### ISSUE TITLE
"Object was never used" errors after upgrading to Tensorflow 1.2

### ISSUE THREAD

**Issue by malcolmsharpe** [openai/pixel-cnn#17]
After upgrading to Tensorflow 1.2, running train.py produces many (567) of the following error:

```
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'model_1/conv2d_0/stack:0' shape=(3,) dtype=int32>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "train.py", line 135, in <module>\n    dropout_p=args.dropout_p, **model_opt)', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/template.py", line 261, in __call__\n    return self._call_func(args, kwargs, check_for_new_variables=True)', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/template.py", line 217, in _call_func\n    result = self._func(*args, **kwargs)', 'File "/media/bruce/MoreData/malcolm/tpc/base-pixel-cnn/pixel_cnn_pp/model.py", line 40, in model_spec\n    x_pad, num_filters=nr_filters, filter_size=[2, 3]))]  # stream for pixels above', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/contrib/framework/python/ops/arg_scope.py", line 181, in func_with_args\n    return func(*args, **current_args)', 'File "/media/bruce/MoreData/malcolm/tpc/base-pixel-cnn/pixel_cnn_pp/nn.py", line 358, in down_shifted_conv2d\n    return conv2d(x, num_filters, filter_size=filter_size, pad=\'VALID\', stride=stride, **kwargs)', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/contrib/framework/python/ops/arg_scope.py", line 181, in func_with_args\n    return func(*args, **current_args)', 'File "/media/bruce/MoreData/malcolm/tpc/base-pixel-cnn/pixel_cnn_pp/nn.py", line 238, in conv2d\n    tf.assert_variables_initialized([V, g, b])', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/tf_should_use.py", line 170, in wrapped\n    return _add_should_use_warning(fn(*args, **kwargs))', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/tf_should_use.py", line 139, in _add_should_use_warning\n    wrapped = TFShouldUseWarningWrapper(x)', 'File "/home/malcolm/data/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/tf_should_use.py", line 96, in __init__\n    stack = [s.strip() for s in traceback.format_stack()]']
==================================
```
It still seems to run OK after printing the errors, though. The culprit seems to be this line, which appears three times in `pixel_cnn_pp/nn.py`:
```
tf.assert_variables_initialized([V, g, b])
```
It seems that this function expects its return value to be used, and that TF 1.2 introduced a check to print an error if the return value isn't used. As a workaround, commenting out this line (in each place it appears) removes the errors.

**Comment by hanss314** [comment]
I am experiencing this too.

**Comment by weixsong** [comment]
me too.

**Comment by ksooklall** [comment]
me three

**Comment by xyc1207** [comment]
me four. Any official solutions?

**Comment by calvinalvin** [comment]
+1

**Comment by tlan16** [comment]
same here

**Comment by JINwonLEE** [comment]
I've got this error too. 
I think you all would better to check that the function which got issue whether deprecated or not.
In my case, the function 'initialize_all_variables' was deprecated in version 1.3.

**Comment by calvinalvin** [comment]
I was able to solve the problem for my particular use case by passing a "shape" for scalar placeholders. So something like this:

```python
learning_rate = tf.placeholder(tf.float32, name='learning_rate', shape=())
keep_prob = tf.placeholder(tf.float32, name='keep_prob', shape=())
```

instead of 

```python
learning_rate = tf.placeholder(tf.float32, name='learning_rate')
keep_prob = tf.placeholder(tf.float32, name='keep_prob')
```

I believe what was happening (in my case), was that I had unit tests making assertions against the shape of these variables but not using them, hence the error. There must have been an extra validation check added for this in tensorflow 1.2

**Comment by luvwinnie** [comment]
I have same issues with this

my placeholder define like this:
<pre>` global_step = tf.get_variable('global_step', [],
                                      initializer = tf.constant_initializer(0),dtype=tf.int32)
        # global_step = tf.Variable(0,name='global_step')
        #global_step=tf.get_variable('global_step', shape=[], initializer=tf.zeros_initializer(), dtype=tf.int32, trainable=False)

        x = tf.placeholder(tf.float32,shape=[None,num_input],name="mfcc_input")
        y = tf.placeholder(tf.int32,shape=[None,num_classes],name="labels")
        is_training = tf.placeholder(dtype=bool, shape=[], name="is_training")
        q_selector = tf.cond(is_training,lambda:[train_mfcc_batch,train_labels_batch],lambda:[test_mfcc_batch,test_labels_batch])`</pre>

and it just shows this error when I try to run on the different machine with GPUs (distributed tensorflow):
<pre>`2017-08-25 12:02:50.917393: W tensorflow/core/framework/op_kernel.cc:1148] Invalid argument: Shape [-1,13] has negative dimensions
2017-08-25 12:02:50.917476: E tensorflow/core/common_runtime/executor.cc:644] Executor failed to create kernel. Invalid argument: Shape [-1,13] has negative dimensions
    [[Node: mfcc_input = Placeholder[dtype=DT_FLOAT, shape=[?,13], _device="/job:worker/replica:0/task:0/gpu:0"]()]]
2017-08-25 12:02:50.918477: W tensorflow/core/framework/op_kernel.cc:1148] Invalid argument: Shape [-1,13] has negative dimensions
2017-08-25 12:02:50.918525: E tensorflow/core/common_runtime/executor.cc:644] Executor failed to create kernel. Invalid argument: Shape [-1,13] has negative dimensions
   [[Node: mfcc_input = Placeholder[dtype=DT_FLOAT, shape=[?,13], _device="/job:worker/replica:0/task:0/gpu:0"]()]]
2017-08-25 12:02:50.920048: W tensorflow/core/framework/op_kernel.cc:1148] Invalid argument: Shape [-1,3] has negative dimensions
2017-08-25 12:02:50.920092: E tensorflow/core/common_runtime/executor.cc:644] Executor failed to create kernel. Invalid argument: Shape [-1,3] has negative dimensions
   [[Node: labels = Placeholder[dtype=DT_INT32, shape=[?,3], _device="/job:worker/replica:0/task:0/gpu:0"]()]]
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'train/report_uninitialized_variables/boolean_mask/Gather:0' shape=(?,) dtype=string>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "sound_classifier_distributed_supervisor.py", line 164, in <module>\n    train_op = rep_op.minimize(xent,global_step=global_step)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/optimizer.py", line 325, in minimize\n    name=name)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/sync_replicas_optimizer.py", line 252, in apply_gradients\n    variables.global_variables())', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 170, in wrapped\n    return _add_should_use_warning(fn(*args, **kwargs))', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 139, in _add_should_use_warning\n    wrapped = TFShouldUseWarningWrapper(x)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 96, in __init__\n    stack = [s.strip() for s in traceback.format_stack()]']`</pre>

**Comment by OranginaGao** [comment]
I met this issue before. What I did to solve this is that I put `shape = []` for `learning_rate` and `keep_prob`

`learning_rate = tf.placeholder(tf.float32, shape=[], name='learning_rate')
  keep_prob = tf.placeholder(tf.float32, shape=[], name='keep_prob')`

**Comment by meteorplus** [comment]
me twelve

**Comment by julypraise** [comment]
me thirteen at tensorflow-gpu 1.11.0

**Comment by korabelnikov** [comment]
me fifteen.
Any suggestions?

**Comment by meteorplus** [comment]
me 6teen

**Comment by meteorplus** [comment]
me 7teen

**Comment by safdar96** [comment]
me 9teen

**Comment by Anton-Velikodnyy** [comment]
me 20teen

**Comment by WolfNiu** [comment]
Me 21! I can drink now. Yay!

**Comment by guillefix** [comment]
Me tuentitu

**Comment by vazyavazya** [comment]
23 =\

**Comment by EricYJA** [comment]
+25

**Comment by aiynmm** [comment]
+26

**Comment by ah3243** [comment]
I think the real problem is that we went from 23 to 25 (failed the counting test) me 24th.

**Comment by hashtala** [comment]
27

**Comment by ntajbakhsh** [comment]
I had a similar issue after upgrading to tf1.12. In my case, the error was thrown at the following line:

tf.assert_equal(tf.shape(x), tf.shape(y)) <-- error
# rest of code

The fix was to place the assert command in a control dependency block, as follows:

with tf.control_dependencies([tf.assert_equal(tf.shape(x), tf.shape(y))]):
      # rest of code

**Comment by vsub21** [comment]
28

**Comment by RomuloUnespian** [comment]
29

**Comment by PrakharPrashantSingh** [comment]
30 -- any one got any solution?

**Comment by jaimix4** [comment]
31 :( I am using tensorflow 1.11

**Comment by PanXiebit** [comment]
32 :(  I am using tensorflow 1.12

**Comment by x5engine** [comment]
thirty three

**Comment by silverriver** [comment]
34, still figuring how to solve it

**Comment by GarlandZhang** [comment]
me 35. This is unrelated to this project (so I'm not sure if this is now general inquiry) but I was having trouble with deploying my Flask API using a tensorflow model onto Heroku with gunicorn. Turns out my Procfile script, which just had `web: gunicorn app:app` was trying to run the main app file which I do not have (before I always just ran `flask run` locally). After adding the app.py file with the usual flask setup, I believe gunicorn actually ran that app using that file, so I was able to fix the problem altogether. Truthfully not sure why it worked though.

**Comment by shubhamb96** [comment]
36

**Comment by LongZhai** [comment]
cannot find pixel_cnn_pp/nn.py in Tensorflow object_detection . what is the full address?

**Comment by amirbalef** [comment]
me 38 

```ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'init_ops/report_uninitialized_variables/boolean_mask/GatherV2:0' shape=(?,) dtype=string>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "train.py", line 184, in <module>
    tf.app.run()  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/deprecation.py", line 324, in new_func
    return func(*args, **kwargs)  File "train.py", line 180, in main
    graph_hook_fn=graph_rewriter_fn)  File "/home/shameon/BillGates/TensorFlow/Codes/models/research/object_detection/legacy/trainer.py", line 416, in train
    saver=saver)  File "/usr/local/lib/python3.6/dist-packages/tensorflow/contrib/slim/python/slim/learning.py", line 791, in train
    should_retry = True  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/util/tf_should_use.py", line 193, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))
==================================```

**Comment by SweeneyLi** [comment]
I solved it!
Originally I use the tensorflow-gpu 1.14.0
Now I use the tensorflow 1.15.0 successfully!

**Comment by yaowang1111** [comment]
I use tf2.1 and still have this problem

**Comment by visva89** [comment]
Tf 2.2 same issue with TensorArray
